janusgraph-core/src/main/java/org/janusgraph/core/JanusGraphLazyRelation.java

```diff
@@ -0,0 +1,264 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.core;
+
+import org.apache.tinkerpop.gremlin.structure.Direction;
+import org.apache.tinkerpop.gremlin.structure.Property;
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+import org.janusgraph.diskstorage.Entry;
+import org.janusgraph.graphdb.internal.InternalRelation;
+import org.janusgraph.graphdb.internal.InternalRelationType;
+import org.janusgraph.graphdb.internal.InternalVertex;
+import org.janusgraph.graphdb.transaction.RelationConstructor;
+import org.janusgraph.graphdb.transaction.StandardJanusGraphTx;
+
+import java.util.Iterator;
+
+public abstract class JanusGraphLazyRelation<V> implements InternalRelation {
+
+    private InternalRelation lazyLoadedRelation;
+    private Entry dataEntry;
+    private final InternalVertex vertex;
+    private final StandardJanusGraphTx tx;
+    private final InternalRelationType type;
+
+    private final Object lockObject;
+
+    public JanusGraphLazyRelation(InternalRelation janusGraphRelation,
+                                  final InternalVertex vertex,
+                                  final StandardJanusGraphTx tx,
+                                  final InternalRelationType type) {
+        this.lazyLoadedRelation = janusGraphRelation;
+        this.dataEntry = null;
+        this.vertex = vertex;
+        this.tx = tx;
+        this.type = type;
+        this.lockObject = new Object();
+    }
+
+    public JanusGraphLazyRelation(Entry dataEntry,
+                                  final InternalVertex vertex,
+                                  final StandardJanusGraphTx tx,
+                                  final InternalRelationType type) {
+        this.lazyLoadedRelation = null;
+        this.dataEntry = dataEntry;
+        this.vertex = vertex;
+        this.tx = tx;
+        this.type = type;
+        this.lockObject = new Object();
+    }
+
+    public InternalRelation loadValue() {
+        if (this.lazyLoadedRelation == null) {
+            synchronized (lockObject) {
+                if (this.lazyLoadedRelation == null) {
+                    if (tx.isClosed()) {
+                        throw new IllegalStateException("Any lazy load operation is not supported when transaction is already closed.");
+                    }
+                    //parseHeaderOnly = false: it doesn't give much latency on read, but gives a good boost on writes
+                    boolean parseHeaderOnly = tx.getConfiguration().isReadOnly();
+                    this.lazyLoadedRelation = RelationConstructor.readRelation(this.vertex, this.dataEntry, parseHeaderOnly, this.tx);
```

Comment:
This doesn't seem right:

> boolean parseHeaderOnly = tx.getConfiguration().isReadOnly();

Reply:
The explanation behind this logic is the following:

By default, when relations are loaded, `parseHeaderOnly` is always `true.`
https://github.com/JanusGraph/janusgraph/blob/ab3a80123844976382dc43acda6d8aaf0ff29183/janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/RelationConstructor.java#L57C27-L57C39.

It is only `false` when you are modifying property or edge, and it copies over the additional properties:
https://github.com/JanusGraph/janusgraph/blob/ab3a80123844976382dc43acda6d8aaf0ff29183/janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/RelationConstructor.java#L40

https://github.com/search?q=repo%3AJanusGraph%2Fjanusgraph%20%22private%20RelationCache%20getPropertyMap%22&type=code

So, for read-only transactions, I keep it `true` as the original implementation, but for writable transactions, I force it to parse headers to avoid the data deserialization being executed twice.

On an extensive dataset, I observed a good improvement by avoiding duplicate deserialization.